PUBDEV-7237 set auto parameters to default values

koniecsveta opened the pull request:
https://0xdata.atlassian.net/projects/PUBDEV/issues/PUBDEV-7237

koniecsveta commented:
> Why do we need categorical encoding = `none`?

We don't need. I was trying to use it with previous approach where I was overwriting parameters directly, but current approach turned out to be better.

koniecsveta commented:
Q: Acording to documentation, For Aggregator, AUTO cat encoding should be One Hot Internal. According to hex.aggregator.Aggregator#init, AUTO cat encoding is Eigen (PUBDEV-3116). Left it like it is - can be changed.

honzasterba commented on h2o-py/h2o/estimators/estimator_base.py:
```diff
@@ -330,6 +330,9 @@ def _resolve_model(self, model_id, model_json):
                 m.parms[p["name"]] = p
+            # build a useful dict of the effective params
+            for p in m._model_json["effective_parameters"]:
+                m.effective_parms[p["name"]] = p    
```
does this code actually do something? it seems to me that m.effective_parms returns a fresh dict so the effect of this code seems to be thrown away? what am I missing?

koniecsveta replied:
this puts effective parameters from model._model_json["effective_parametes"] to model.effective_parms with all the attributes they have ('_meta', 'name', 'label', 'help', 'required', 'type', 'value' ...). The only difference is, that you can get specific parameter like xgb1.**effective_parms**['fold_assignment'] instead of xgb1._model_json['effective_parameters'][8] - what is a bit nicer for subsequent usage. Then in model_base.py, by model.effective_params you get dict {nameOfParameter1 : valueOfParameter1, nameOfParameter2 : valueOfParameter2, ...} so you can get value of fold assignment parameter of xgb1 model like xgb1.**effective_params**['fold_assignment']. Did it like this in consistency with model._model_json['parameters'], model.parms and model.actual_params. Do you think that we shouldn't have model.effective_parms and I should extract model.effective_params directly from model_json["effective_parameters"]?

honzasterba replied:
oh got it, got confused the field effective_parms and @property def effective_params

honzasterba commented on h2o-algos/src/main/java/hex/deeplearning/DeepLearning.java:
```diff
@@ -326,6 +327,36 @@ public final void buildModel() {
       }
     }
+    
+    void initDefaultParam(DeepLearningModel deepLearningModel) {
```
shoud be probably called initEffectiveParam

honzasterba commented on h2o-algos/src/main/java/hex/glm/GLMModel.java:
```diff
@@ -40,6 +41,16 @@ public GLMModel(Key selfKey, GLMParameters parms, GLM job, double [] ymu, double
   }
 
+  void initDefaultParam() {
+    if (_parms._fold_assignment == Model.Parameters.FoldAssignmentScheme.AUTO) {
```
this code is repeated over and over, maybe moved it to Model.class or some helper class:

koniecsveta replied:
moved to EffectiveParametersUtils

honzasterba commented on h2o-algos/src/main/java/hex/kmeans/KMeansModel.java:
```diff
@@ -70,7 +67,20 @@
+  }
+
+  void initDefaultParam() {
+    if (_parms._fold_assignment == Model.Parameters.FoldAssignmentScheme.AUTO) {
```
also repeated code

michalkurka commented:
Looks good! What I am missing is
- how do we make sure that adding a new "AUTO" parameter will be reflected in change to the code that populates "Effective Parameters"

Can we automagically detect all "AUTO" parameters (at some point) and check that the effective parameters do not have any AUTO - eg. put a generic assert somewhere?

honzasterba commented on h2o-core/src/main/java/hex/ModelBuilder.java:
```diff
@@ -264,6 +265,20 @@ public boolean onExceptionalCompletion(Throwable ex, CountedCompleter caller) {
     public abstract void computeImpl();
   }
 
+  public void assertEffectiveParameters(Model.Parameters effectiveParameters){
```
needs a better name - my idea: checkEffectiveParmsDoesNotContainAuto

koniecsveta replied:
renamed

honzasterba commented on h2o-core/src/main/java/hex/ModelBuilder.java:
```diff
@@ -264,6 +265,20 @@ public boolean onExceptionalCompletion(Throwable ex, CountedCompleter caller) {
+      for (Field field : effectiveParameters.getClass().getFields()) {
+        Class type = field.getType();
+        Object value = field.get(effectiveParameters);
+        if (value != null/* && "AUTO".equals(value.toString())*/){
```
commented out code

koniecsveta replied:
fixed

honzasterba commented on h2o-core/src/main/java/hex/ModelBuilder.java:
```diff
@@ -264,6 +265,20 @@ public boolean onExceptionalCompletion(Throwable ex, CountedCompleter caller) {
+        Class type = field.getType();
+        Object value = field.get(effectiveParameters);
+        if (value != null/* && "AUTO".equals(value.toString())*/){
+          assert(!"AUTO".equals(value.toString())) : "FAIL: found AUTO value in effective parameters: " + field.getName();
```
remove the "FAIL: " the assertion error sais that already

honzasterba replied:
maybe use equalsIgnoreCase?

honzasterba commented on h2o-core/src/main/java/hex/ModelBuilder.java:
```diff
@@ -264,6 +265,20 @@ public boolean onExceptionalCompletion(Throwable ex, CountedCompleter caller) {
+        }
+      }
+    } catch (IllegalAccessException e) {
+      e.printStackTrace();
```
Params should have all fields public/readable so this should not happen so rethrowing as RuntimeException should be fine

koniecsveta replied:
fixed

honzasterba commented on h2o-algos/src/main/java/hex/aggregator/Aggregator.java:
```diff
@@ -231,6 +230,7 @@ public void computeImpl() {
         }
         _job.update(1, "Done.");
         model.update(_job);
+        checkEffectiveParmsDoesNotContainAuto(model._effective_parms);
```
this is also error-prone and could be simplified. I think we shoud ideally move this call to Model so that we do not need duplicate it in every model sub-class
can you think of a good place where to have this? is it after train or after init? 
ideally I would do the same for initEffectiveParams() maybe make that an overridable method in Model and call it at some defined place, also how about renaming that to computeEffectiveParams, to make it more clear?

koniecsveta replied:
initEffectiveParams() will be easy. checkEffectiveParmsDoesNotContainAuto() should be in the end/ after hex.ModelBuilder.Driver#computeImpl(). can try to find some place for it

honzasterba commented on h2o-core/src/main/java/hex/ModelBuilder.java:
```diff
@@ -262,6 +264,29 @@ public boolean onExceptionalCompletion(Throwable ex, CountedCompleter caller) {
+        }
+      }
+    } catch (IllegalAccessException e) {
+      throw new RuntimeException(e);
```
I would add message exaplaining when the error happened "Error while checking params for auto values"

koniecsveta replied:
added

honzasterba commented on h2o-algos/src/main/java/hex/aggregator/AggregatorModel.java:
```diff
@@ -56,8 +56,16 @@ public ToEigenVec getToEigenVec() {
+    super(selfKey,parms,output);
+  }
 
+  public void initEffectiveParam() {
```
missing @override

koniecsveta replied:
fixed

honzasterba commented on h2o-core/src/main/java/hex/ModelBuilder.java:
```diff
@@ -262,6 +264,29 @@ public boolean onExceptionalCompletion(Throwable ex, CountedCompleter caller) {
+      M model = _result.get();
+      if (model != null) {
+        model._effective_parms = (P) _parms.clone();
+        model.initEffectiveParam();
```
I am not sure I like the computeEffectiveParameters/initEffectiveParam naming, I think both could be the same.

koniecsveta replied:
renamed to be computeEffectiveParameters both

honzasterba commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoost.java:
```diff
@@ -396,6 +396,24 @@ final void buildModelImpl() {
     }
+    
+    @Override
+    public void computeEffectiveParameters() {
```
why is this not done in model.initEffectiveParams? overriding this method seems like a not good idea, ideally I would make the super method final, becuase overriding is dangerous (might forget to add checkEffectiveParmsDoesNotContainAuto for example)

koniecsveta replied:
this is done here because I have here isTrainDatasetSparse() and train().numRows()

koniecsveta replied:
moved to model

honzasterba commented on h2o-algos/src/main/java/hex/kmeans/KMeans.java:
```diff
@@ -438,6 +439,15 @@ public void computeImpl() {
     }
+    
+    @Override
+    public void computeEffectiveParameters() {
```
any reason this cant be in model class?

koniecsveta replied:
originally I was calling model.initEffectiveParam() in the constructor so _output was not available yet and cases where output was needed I was solving separately. how the solution evolved I missed that I have it available at the moment when I call model.initEffectiveParam now. So all this can be now moved to model class. that's also the reason why it is called init... but now it can be called compute... will fix

koniecsveta replied:
fixed

honzasterba commented on h2o-algos/src/main/java/hex/deeplearning/DeepLearning.java:
```diff
@@ -326,6 +338,13 @@ public final void buildModel() {
       }
     }
+    
+    void initEffectiveParam(DeepLearningModel deepLearningModel) {
```
why not in model, here the orignal model. initEffectiveParam is not even ever called

koniecsveta replied:
you're right, this whole initEffectiveParam was to be deleted and and original model.initEffectiveParam was to be called in computeEffectiveParameters() on line 224. Thanks!

koniecsveta replied:
fixed

honzasterba commented on h2o-core/src/main/java/hex/ModelBuilder.java:
```diff
@@ -262,6 +264,29 @@ public boolean onExceptionalCompletion(Throwable ex, CountedCompleter caller) {
+    public void computeEffectiveParameters() {
+      M model = _result.get();
+      if (model != null) {
+        model._effective_parms = (P) _parms.clone();
```
is this necessary, clone() is called in model constructor

koniecsveta replied:
its not, removed

honzasterba commented on h2o-core/src/main/java/hex/ModelBuilder.java:
```diff
@@ -262,6 +264,29 @@ public boolean onExceptionalCompletion(Throwable ex, CountedCompleter caller) {
 
     public abstract void computeImpl();
+    
+    public void computeEffectiveParameters() {
```
ideally I would make this method final

koniecsveta replied:
fixed

honzasterba commented on h2o-algos/src/main/java/hex/deeplearning/DeepLearningModel.java:
```diff
@@ -2217,6 +2226,7 @@ static void modifyParms(DeepLearningParameters fromParms, DeepLearningParameters
         // Automatically set the distribution
         if (fromParms._distribution == DistributionFamily.AUTO) {
           // For classification, allow AUTO/bernoulli/multinomial with losses CrossEntropy/Quadratic/Huber/Absolute
+          if (nClasses == 1) toParms._distribution = DistributionFamily.gaussian;
```
is this necessary? the else branch seems to handle this case

koniecsveta replied:
its wrong, thanks

honzasterba commented on h2o-algos/src/main/java/hex/deeplearning/DeepLearningModel.java:
```diff
@@ -72,6 +73,14 @@ public DeepLearningModelOutput(DeepLearning b) {
   } // DeepLearningModelOutput
 
+    @Override
+    public void computeEffectiveParameters() {
```
should this also call super.?

koniecsveta replied:
it could, but I think its not necessary. by calling super.computeEffectiveParameters() firstly, it would set distribution, stopping_metric, fold_assignment and categorical encoding to null, and then after super. set them to effective values by respective logic. there is assert(_dist._family != DistributionFamily.AUTO); in both DeepLearningModel constructors, so it wont let effective distribution = AUTO, and in other lines it also cannot happen that it will let some of those effective parameters = AUTO.

honzasterba commented on h2o-algos/src/main/java/hex/kmeans/KMeansModel.java:
```diff
@@ -70,7 +68,17 @@
+
+  @Override
+  public void computeEffectiveParameters() {
+    _effective_parms._distribution = null;
```
super?

koniecsveta replied:
similarly, super would set distribution, stopping_metric, fold_assignment and categorical encoding to null, and then the rest after super would set fold_assignment and categorical encoding to effective value. distribution and stopping_metric are not used in KMeans, they are not returned on clients as parametes, but in this stage we have them present. hence they are set to null. Otherwise, they would be = AUTO and wouldn't pass through checkEffectiveParmsDoesNotContainAuto()

honzasterba commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java:
```diff
@@ -222,6 +223,25 @@ public XGBoostModel(Key<XGBoostModel> selfKey, XGBoostParameters parms, XGBoostO
+    EffectiveParametersUtils.initFoldAssignment(_parms, _effective_parms);
+    EffectiveParametersUtils.initDistribution(_parms, _effective_parms, _output.nclasses());
+    _effective_parms._backend = getActualBackend(_parms);
+    _effective_parms._dmatrix_type = _output._sparse ? XGBoostModel.XGBoostParameters.DMatrixType.sparse : XGBoostModel.XGBoostParameters.DMatrixType.dense;  
```
this looks like a copy paste from XGBoost.java could we move the methods here rather and call them from here?

honzasterba commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java:
```diff
@@ -222,6 +223,27 @@ public XGBoostModel(Key<XGBoostModel> selfKey, XGBoostParameters parms, XGBoostO
+    EffectiveParametersUtils.initDistribution(_parms, _effective_parms, _output.nclasses());
+    _effective_parms._backend = getActualBackend(_parms);
+    _effective_parms._dmatrix_type = _output._sparse ? XGBoostModel.XGBoostParameters.DMatrixType.sparse : XGBoostModel.XGBoostParameters.DMatrixType.dense;
+    //tree_method parameter is evaluated according to https://github.com/h2oai/xgboost/blob/96f61fb3be8c4fa0e160dd6e82677dfd96a5a9a1/src/gbm/gbtree.cc#L127 + we don't 
```
missing space

honzasterba commented on h2o-py/tests/testdir_jira/pyunit_pubdev_7237_get_actual_auto_parameter.py:
```diff
@@ -0,0 +1,240 @@
+
+
+def test_set_auto_parameters_to_default_values():
+    #testing default setup of following parameters:
```
I would split this test into separate tests per algo, will be easier to maintain

honzasterba commented on h2o-algos/src/main/java/hex/aggregator/AggregatorModel.java:
```diff
@@ -56,8 +56,15 @@ public ToEigenVec getToEigenVec() {
+  }
 
+  @Override
+  public void computeEffectiveParameters() {
```
unnecessary

honzasterba commented on h2o-py/tests/testdir_jira/pyunit_pubdev_7237_effective_parameters.py:
```diff
@@ -0,0 +1,268 @@
+import sys
```
I think testdir_jira is more for is more for bugfixes, this is a new feature, my original suggestion was to split this into separate tests for each algo in testdir_algos, but I see that for your dev this is better because you want to test all the algos at once, but in the future you would usually change only one algo and want to touch only the test for that algo

honzasterba replied:
I also think we should add atleast a simple test for R

honzasterba commented on h2o-r/h2o-package/R/kvstore.R:
```diff
@@ -176,63 +176,94 @@ h2o.getModel <- function(model_id) {
+  effective_parameters <-list()
+  effective_allparams <- list()
+    
+    fill_parameters <- function(param, effective=FALSE) {
```
something has wrong indent here

honzasterba commented on h2o-r/h2o-package/R/kvstore.R:
```diff
@@ -236,13 +239,26 @@ h2o.getModel <- function(model_id) {
-      allparams[[name]] <<- value
+        
+      if (effective)
+        effective_allparams[[name]] <<-value
```
having this function modify its outside scope is not very good design, also <<- operator should not be used
I think using llply should do the trick

honzasterba commented on h2o-r/h2o-package/R/kvstore.R:
```diff
@@ -236,13 +239,40 @@ h2o.getModel <- function(model_id) {
+  }
+    
+  # get name, value pairs
+  allparams_key_val = llply(json$parameters, fill_pairs, all=TRUE)
```
it seems we do not depend on plyr :( so this will probably not be possible without adding a dependency
@michalkurka WDYT?

michalkurka replied:
good catch, thank you @honzasterba 

yes, the code needs to work without any additional dependencies

koniecsveta replied:
so can I use the previous approach with <<- in this case?

honzasterba commented on h2o-r/h2o-package/R/kvstore.R:
```diff
@@ -210,8 +210,11 @@ h2o.getModel <- function(model_id) {
   allparams  <- list()
-  lapply(json$parameters, function(param) {
-    if (!is.null(param$actual_value)) {
+  effective_parameters <-list()
```
whitespace

honzasterba commented on h2o-r/h2o-package/R/kvstore.R:
```diff
@@ -236,13 +239,38 @@ h2o.getModel <- function(model_id) {
+  for (param in effective_allparams_key_val) {if (!any(is.na(param[1])))  effective_allparams[unlist(param[1])] <- param[2]}
+  for (param in parameters_key_val) {if (!any(is.na(param[1]))) parameters[unlist(param[1])] <- param[2]}
+  for (param in effective_parameters_key_val) {if (!any(is.na(param[1]))) effective_parameters[unlist(param[1])] <- param[2]}
+
```
too many empty lines

honzasterba commented on h2o-algos/src/main/java/hex/gam/GAM.java:
```diff
@@ -494,6 +491,7 @@ private void copyGLMtoGAMModel(GAMModel model, GLMModel glmModel) {
       model._rank = glmModel._output.bestSubmodel().rank();
       model._ymu = new double[glmModel._ymu.length];
       System.arraycopy(glmModel._ymu, 0, model._ymu, 0, glmModel._ymu.length);
+      model._parms = copyGLMParams2GAMParams(glmModel._effective_parms);
```
should we really change _parms here? I think the point was to leave _parms as is, so this should probably copy to _effective_parms, no?

koniecsveta replied:
If I copy to effective params, it won't work because param.solver is set during GLM training. after training it is copied to effective_param.solver in hex.ModelBuilder.Driver#computeEffectiveParameters. when I create GAM, I have solver = AUTO and its not set then nowhere so after GAM is trained and I will get to hex.ModelBuilder.Driver#computeEffectiveParameters I will be copying param.solver = AUTO to effective_prams.solver and end up with AUTO in effective parameters.. maybe I can do `model._parms = copyGLMParams2GAMParams(glmModel._parms);` like this effective params wouldn't interact with params. Or you think GLM params shouldnt interact with GAM params at all here?

koniecsveta commented:
1. To fix failing .R generic tests  during compare_params(original_model, generic_model) I've changed compare_params() so that if compares values from the same source as it was taking before. 

it was like this:
model@parameters taken from [json][output][parameters]
model@allparameters taken from [json][output][parameters]

parameters subset of allparameters

now it is like this:
model@parameters taken from [json][output][effective_parameters]
model@allparameters taken from [json][output][parameters]

so now it compares the same subset (for cycle iterating through @parameters) of them from the same source (taking them from @allparameters)
2. Changed h2o::getParms so that its taken from the same source as it was before (so taking them from @allparameters and not from @parameters)


3. there will need to be more fixes in runit_stackedensemble_metalearner.R and runit_stackensemble_nfolds.R because of the fact, that model@parameters changed its content, for example: expect_equal(stack0@parameters$metalearner_algorithm, NULL), but now it retruns effective params so its already set to "glm". Dont know whether I am supposed to touch this code?

honzasterba commented:
since the semantics of model@parameters has changed I think its OK to fix the failing tests to match the current semantics

sebhrusen commented:
I'm still confused, and I believe that many users will be even more confused.
So, at schema level we keep 2 sets of parameters: `parameters` and `effectiveParameters`.
- `effectiveParameters` holds a value before the training but after converting AUTO to "actual" value.
- `parameters` still holds the same legacy value without any proper semantic as this value may have been set by user, or dynamically when initializing the driver, or even after cross-validation and before training final model...
- Both sets also hold the default value.

From there I have a few concerns:
- legacy `parameters` is still half-buggy as it doesn't really contain the value set by the user but a mix of uninterpretable things.
- from a  REST perspective we have "model.parameters.foo.actual_value" and "model.effectiveParameters.foo.actual_value" which is pretty confusing.
- in `Model`, `_effective_parms` seem to be completely dissociated from the training logic which is actually still using `_parms`, and is computed independently, which means that it is very easy to forget something there or worse have a value that is not even the actual one, typically after adding a new param.

Suggestion: can't we, instead of adding `effectiveParameters` to the schema, extend `ModelParameterSchemaV3`? This would keep a single entry-point for the parameters, removing part of the confusion, and also the duplicate defaults.
Ideally, there, we would have `default_value` , `input_value`, `actual_value`. No one is interested in any other combination like we find sometimes in legacy `actual_value`.

My feeling (but maybe I'm wrong) is that it would have been much simpler to take a snaphot of `_parms` before the training and ensure that this value is then exposed as the `input_value`.
For the `actual_value` , instead of relying on `_effective_parms`, why not simply change the `_parms` value during the model initialization, this way we guarantee that this will really be the actual value used during training, e.g. in your utility functions:
```
    public static void initFoldAssignment(
        Model.Parameters params,
    ) {
        if (params._fold_assignment == Model.Parameters.FoldAssignmentScheme.AUTO) {
            if (params._nfolds > 0 && params._fold_column == null){
                params._fold_assignment = Model.Parameters.FoldAssignmentScheme.Random;
            } else {
                params._fold_assignment = null;
            }
        }
    }
```

This way it resolves most of our parameters issues:
- `default_value` remains what it is.
- `input_value` comes from the snapshot of params before init, so it contains defaults overridden with user-provided params, those values allow to reproduce a model even when using cross-validation.
- `actual_value` is the value of `_parms` at the end of training, and as such it is expected to be the true actual value for the final model.

Any thoughts?

honzasterba commented:
I like what @sebhrusen is saying. We could even go as far as to change the lifecycle to copy-input-parms -> compute-effective-parms -> train
This way we would avoid the code duplication that we have node between train & compute effective params. @michalkurka WDYT?

sebhrusen commented on h2o-core/src/main/java/water/api/schemas3/ModelSchemaV3.java:
```diff
@@ -34,6 +34,9 @@
 
+  @API(help="The input parameters for the model (what was actually used, when parameter was not specified/was specified as AUTO).", direction=API.Direction.OUTPUT)
+  private PS input_parameters;
+  
```
is it still needed now that `ModelParameterSchemaV3` has an `input_value` field?

koniecsveta replied:
no no, not anymore - I just wanted to make work input_values first and after that get rid of effective parameters

sebhrusen replied:
@koniecsveta replying to https://github.com/h2oai/h2o-3/pull/4361#issuecomment-643996065

I just want to ensure we don't expose twice the same thing or worse, make it confusing.
I just realize that you made it private, so it is probably not exposed in the REST API: can you please remove the `@API` annotation then?
It should just be needed for this call below, right?
```
      ModelParametersSchemaV3.writeParametersJSON(ab, parameters, input_parameters, defaults ,"parameters");
```

honzasterba commented on h2o-algos/src/main/java/hex/deeplearning/DeepLearning.java:
```diff
@@ -203,14 +206,33 @@ protected int desiredChunks(final Frame original_fr, boolean local) {
       //check that _parms isn't changed during DL model training
-      long cs2 = _parms.checksum();
-      assert(cs == cs2);
+      Model.Parameters parmsToCheck2 = _parms.clone();
```
why do we need this clone, does checkNonAutoParmsNotChanged modify any of its arguments?

koniecsveta replied:
before, it was checking all the parameters to not to be changed. Now I need to evaluate/change distribution parameter in case it is AUTO to its actual value during buildModel(), because I can't tell it beforehand, it depends on nClasses and loss function and I set it after the place where it is evaluated and set to model_info inside buildModel() inside DeepLearningParameters.Sanity.modifyParms(_parms, cp.model_info().get_params(), nclasses()); (DeepLearning.javaL312)
Hence I rewrote this check: it assumes that parameters with value AUTO will be evaluated during buildModel() (which is only distribution parameter) and checks for not being changed all the others (instead checking all the parameters).

honzasterba commented on h2o-algos/src/main/java/hex/kmeans/KMeans.java:
```diff
@@ -1,6 +1,7 @@
 package hex.kmeans;
 
 import hex.*;
+import hex.util.EffectiveParametersUtils;
```
unnecesary import?

honzasterba commented on h2o-algos/src/test/java/hex/glm/GLMTest.java:
```diff
@@ -760,6 +763,7 @@ public void testBounds() {
       params._alpha = new double[]{1};
       params._lambda = new double[]{0.001607};
       params._obj_reg = 1.0/380;
+      params._fold_assignment = Model.Parameters.FoldAssignmentScheme.AUTO;
```
why are you adding this auto everywhere?

koniecsveta replied:
because fold assignment is only applicable if a value for nfolds is specified and a fold_column is not specified. Otherwise it is evaluated as null. if you then reuse parameters like this (with fold_assignment = null)for next model building, it creates a problem, because you want to have it set to AUTO (not null) in input so that it can be evaluated in next model computation.

koniecsveta replied:
enabled null as fold assignment input value

honzasterba commented on h2o-core/src/main/java/hex/ModelBuilder.java:
```diff
@@ -191,12 +195,15 @@ private static int ensureBuilderIndex(String urlName) {
   /** All the parameters required to build the model. */
   public P _parms;              // Not final, so CV can set-after-clone
 
+  /** All the parameters required to build the model. */
```
JavaDoc needs updating

honzasterba commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java:
```diff
@@ -229,6 +230,61 @@ public XGBoostModel(Key<XGBoostModel> selfKey, XGBoostParameters parms, XGBoostO
+    _parms._dmatrix_type = _output._sparse ? XGBoostModel.XGBoostParameters.DMatrixType.sparse : XGBoostModel.XGBoostParameters.DMatrixType.dense;
+  }
+  
+  // useful for debugging
```
was removed by a different change, please do no add back

koniecsveta replied:
I must have overseen it during conflicts resolving.

honzasterba commented on h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java:
```diff
@@ -839,6 +839,7 @@ public void testBinomialTrainingWeights() {
       DKV.put(trainingFrameSubset);
       parms._weights_column = null;
       parms._train = trainingFrameSubset._key;
+      parms._fold_assignment = Model.Parameters.FoldAssignmentScheme.AUTO;
```
can be removed?

koniecsveta commented on h2o-algos/src/main/java/hex/ensemble/StackedEnsembleModel.java:
```diff
@@ -42,6 +42,14 @@ public StackedEnsembleModel(Key selfKey, StackedEnsembleParameters parms, Stacke
+  @Override
+  public void initActualParamValues() {
+    super.initActualParamValues();
+    if (_parms._metalearner_fold_assignment == AUTO) {
```
there is _params.evaluate_auto check on place where super method is being called. (Model.java#Model()L958)

koniecsveta commented on h2o-algos/src/main/java/hex/tree/isofor/IsolationForestModel.java:
```diff
@@ -54,7 +54,24 @@ public ModelCategory getModelCategory() {
+  @Override
+  public void initActualParamValues() {
+    super.initActualParamValues();
+    if (_parms._stopping_metric == ScoreKeeper.StoppingMetric.AUTO){
```
it is checked before initActualParamValues() is called as described above.

koniecsveta commented:
https://github.com/h2oai/h2o-3/pull/4361#discussion_r438350094
Here I thought you commented previous field effective_parms, which was removed. This is a new field with different purpose. input_params is used to store information about input state of parameters. Previously, you suggested extension of ModelParameterSchemaV3 to have default_value , input_value, actual_value. (here https://github.com/h2oai/h2o-3/pull/4361#issuecomment-617878703) Do you think, that input state of parameters should be handled in a different way?

koniecsveta commented on h2o-algos/src/main/java/hex/ensemble/StackedEnsemble.java:
```diff
@@ -328,6 +328,8 @@ public void computeImpl() {
                 " but must be one of " + Arrays.toString(Metalearner.Algorithm.values()));
       }
+      if (_model._parms._evaluate_auto && _model._parms._metalearner_algorithm == Metalearner.Algorithm.AUTO)
+        _model._parms._metalearner_algorithm = Metalearner.Algorithm.glm;
```
fixed

koniecsveta commented on h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java:
```diff
@@ -889,6 +889,8 @@ public void testRegressionTrainingWeights() {
       parms._weights_column = null;
       parms._train = trainingFrameSubset._key;
+      parms._categorical_encoding = Model.Parameters.CategoricalEncodingScheme.AUTO;
+      parms._fold_assignment = Model.Parameters.FoldAssignmentScheme.AUTO;
```
it works also unset, removed

koniecsveta commented on h2o-py/tests/testdir_algos/automl/pyunit_automl_leaderboard.py:
```diff
@@ -239,7 +239,7 @@ def test_AUTO_stopping_metric_with_no_sorting_metric_binomial():
     first = [m for m in non_se if 'XGBoost_1' in m]
     others = [m for m in non_se if m not in first]
-    check_model_property(first, 'stopping_metric', True, "AUTO")
+    check_model_property(first, 'stopping_metric', True, None) #if stopping_rounds == 0, actual value of stopping_metric is set to None
```
yes, but I think there is exception in this for XGBoost, isn't it? (AutoMLTest.java#testAlgosHaveDefaultParametersEnforcingReproducibility()L713)

sebhrusen replied:
Thanks for pointing me to that test. However, XGBoost should not handle `stopping_rounds` differently anymore.
Just created https://0xdata.atlassian.net/browse/PUBDEV-7618 to clean the java test.
Can you please run the test without your change?
If it's passing, then please keep the old code.
If it fails, then just mention `PUBDEV-7618` in your comment: I'll look at it with that ticket, thx.

koniecsveta replied:
as discussed, defaulr stopping_rounds is 3 but when CV is used, the final model resets it to 0 in cv_computeAndSetOptimalParameters. But fixed input params saving for AutoML in
e8f39bb

koniecsveta commented on h2o-algos/src/main/java/hex/util/EffectiveParametersUtils.java:
```diff
@@ -0,0 +1,68 @@
+import hex.tree.SharedTreeModel;
+
+
+public class EffectiveParametersUtils {
```
I don't think it should be checked here. It works like if (evaluate_auto) {initActualParamValues()} and in initActualParamValues you set values either by some method from EffectiveParametersUtils.java or by actual commands if it is something model specific.

params._evaluate_auto is a safety net. Switcher to be able to use the before-this-change version of param evaluation easily, in case we find out something is being set incorrectly and it messes up calculation. Maybe it really could be done on some different level. Do you know whether we have some similar switcher so that it could be done in a similar pattern?

sebhrusen replied:
In that case, I think it makes more sense to enable/disable this globally as a system property.
See `SQLManager.shouldUseTempTable()` or `automl.Algo.enabled()` for examples.
Just be careful to never store the value of this system property in a class field or in a singleton: user should be able to change the value of this property dynamically without having to restart the backend, that's why I was also suggesting the client `backend property` API to switch those easily.

koniecsveta replied:
f76ba39

koniecsveta commented on h2o-core/src/main/java/hex/ModelBuilder.java:
```diff
@@ -273,6 +281,18 @@ public boolean onExceptionalCompletion(Throwable ex, CountedCompleter caller) {
+      if (model != null) {
+        //set input parameters
+        model.setInputParms(_input_parms);
+        //set _fold_assignment param to null in case it was entered as AUTO and it wasn't set during the training
```
I can't remember other reason I did this like this than because fold_assignment didn't have null enabled as input value. Now as it has it enabled I can remove this and set fold assignment to null in EffectiveParametersUtils#initFoldAssignment() in else {...}. Logically it belongs there but it wasn't possible with null disabled.

koniecsveta replied:
fixed

koniecsveta commented on h2o-core/src/main/java/hex/ModelBuilder.java:
```diff
@@ -1214,10 +1237,13 @@ public void init(boolean expensive) {
       hide("_keep_cross_validation_fold_assignment", "Only for cross-validation.");
       hide("_fold_assignment", "Only for cross-validation.");
-      if (_parms._fold_assignment != Model.Parameters.FoldAssignmentScheme.AUTO) {
+      if (_parms._fold_assignment != Model.Parameters.FoldAssignmentScheme.AUTO & _parms._fold_assignment != null) {
```
fixed

koniecsveta commented on h2o-algos/src/test/java/hex/deeplearning/DeepLearningAutoEncoderTest.java:
```diff
@@ -64,11 +66,12 @@ public void run() {
-        p._standardize = false;
-        DeepLearning dlNoStand = new DeepLearning(p);
+        p2._standardize = false;
+        DeepLearning dlNoStand = new DeepLearning(p2);
```
Here, stopping_metric got evaluated. I can reuse p and specify stopping_metric = AUTO.

koniecsveta replied:
3cd7357

honzasterba commented on h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java:
```diff
@@ -229,6 +230,37 @@ public XGBoostModel(Key<XGBoostModel> selfKey, XGBoostParameters parms, XGBoostO
+    // tree_method parameter is evaluated according to:
+    // https://github.com/h2oai/xgboost/blob/96f61fb3be8c4fa0e160dd6e82677dfd96a5a9a1/src/gbm/gbtree.cc#L127 
+    // + we don't use external-memory data matrix feature in h2o 
+    // + https://github.com/h2oai/h2o-3/blob/b68e544d8dac3c5c0ed16759e6bf7e8288573ab5/h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java#L348
```
I agree that the duplication is no ideal, maybe we can refactor this into a method that would be called from here and also from createParamsMap?

koniecsveta replied:
ba8b14e

sebhrusen commented on h2o-bindings/bin/bindings.py:
```diff
@@ -360,23 +360,30 @@ def find_field(fields, field_name):
             schema = m[base]
             schema["generics"] = generics
-            generic_map = {long_type: gen_type for gen_type, long_type in generics}
+            generic_map = DictList()
```
besides, with `default_dict` you see immediately what kind of structure you're manipulating:
```
generic_map[schema_name].append(name)
```
instead of
```
generic_map[schema_name] = name
```
the latter requiring you to know that the datastructure overrides `__setitem__`, which should be done only when "necessary"

koniecsveta replied:
f75f55b

sebhrusen commented on h2o-py/h2o/h2o.py:
```diff
@@ -2211,6 +2211,20 @@ def print_mojo(mojo_path, format="json", tree_index=None):
+
+    :example:
+
+    >>> h2o.set_system_property("sys.ai.h2o.algos.evaluate_auto_model_parameters", "false")
```
I'm not sure we should provide direct access to all system properties like that, I would rather limit to some feature flags for now: 
1. there may be security concerns.
2. those properties are not discoverable by the user.
3. validation concern: if user makes a typo in the property name, the property will be set and user won't notice.

My suggestion:
- backend always adds the "sys.ai.h2o." prefix. This will strictly limit the kind of properties that the clients can change (it may still be too broad).
- expose a map on the client with just the flags we want to expose so that the user knows what can be switched on/off.

 I would limit this to feature flags for now, for example for the Py client
```
feature_flags = dict(
   enable_evaluation_of_auto_model_parameters = ("algos.evaluate_auto_model_parameters", True),
   enable_xgboost_in_automl_multinode = ("automl.xgboost.multinode.enabled", True)
)

def set_feature_flag(feature, enable=None):
   assert feature in feature_flags
   key, def_value = feature_flags[feature]
   value = def_value if enable is None else enable
   value = "true" if value else "false"
   api("POST /3/SetSystemProperty", data=dict(property=key, value=value))
```

Now user can do 
```
h2o.set_feature_flag("enable_evaluation_of_auto_model_parameters", False)
And it will fail if there's a typo in the property.
```

I also have a concern with the name: I think we should specify that those properties or feature flags are only for the backend, otherwise it may confuse some users as some client features can also be switched using the `h2oconfig` file.
My suggestion would be to expose this only in the `h2o.backend` module for clarity.

Also, there may be concerns that I didn't notice yet (@michalkurka ?), so I wouldn't mind if you prefer to implement this in a separate PR.

koniecsveta replied:
Hm yes, this approach with `assert feature in feature_flags` will give us much better control over what is being set by this. With separate PR... I think that the switcher of the new feature should be a part of the PR with the new feature, so I would do this here at least in some very specific form (only `enable_evaluation_of_auto_model_parameters `flag in feature_flags). If some concerns will raise afterwards, it can still be perfected/extended for other features/etc in subsequent PR

koniecsveta replied:
6fc83f2

koniecsveta commented on h2o-core/src/main/java/hex/Model.java:
```diff
@@ -1011,12 +1017,25 @@ public Model(Key<M> selfKey, P parms, O output) {
+   */
+  public static boolean evaluateAutoModelParameters() {
+      return Boolean.parseBoolean(System.getProperty(EVAL_AUTO_PARAMS_ENABLED, "true"));
+  }
```
It doesn't seem like we abide by this convention but at least moved evaluateAutoModelParameters() up under the first static method of the class

koniecsveta commented on h2o-core/src/main/java/water/api/schemas3/SetSystemPropertyV3.java:
```diff
@@ -0,0 +1,15 @@
+  public String property = "";
+
+  @API(help = "Factor columns", direction=API.Direction.INPUT)
+  public boolean value = true;
```
renamed/fixed, thx

koniecsveta commented on h2o-algos/src/test/java/hex/ensemble/StackedEnsembleTest.java:
```diff
@@ -378,7 +378,7 @@ public void testBaseModelPredictionsCaching() {
         seParams._base_models = grid.getModelKeys();
         seParams._seed = seed;
-
+        StackedEnsembleParameters seParamsInitialClone = (StackedEnsembleParameters) seParams.clone();
```
7e81e9d

koniecsveta commented on h2o-core/src/main/java/water/api/schemas3/ModelParameterSchemaV3.java:
```diff
@@ -152,6 +160,15 @@ public final AutoBuffer writeJSON_impl(AutoBuffer ab) {
+    } else {
+      ab.putJSONStr("input_value").put1(':').putJSON(input_value);
+      ab.put1(',');
+    }
```
180a54d

koniecsveta commented on h2o-core/src/main/java/hex/Model.java:
```diff
@@ -999,33 +1010,37 @@ public ScoringInfo last_scored() {
-      default:
-        return (float) (_output.isClassifier() ? logloss() : _output.isAutoencoder() ? mse() : deviance());
-
+    if (_parms._stopping_metric == null) {
```
c8e6384

sebhrusen commented on h2o-bindings/bin/bindings.py:
```diff
@@ -368,15 +368,22 @@ def find_field(fields, field_name):
+                        gen_type.pop(0)
+                    else:    
+                        field["schema_name"] = gen_type[0]
+                        mapped_fields[field["name"]] = generic_index[gen_type[0]]
```
looks like only `gen_type.pop(0)` should be in the `if` condition, the other statements are always executed.

sebhrusen commented on h2o-core/src/main/java/hex/Model.java:
```diff
@@ -1011,12 +1024,19 @@ public Model(Key<M> selfKey, P parms, O output) {
     super(selfKey);
     assert parms != null;
     _parms = parms;
+    evalAutoParamsEnabled = evaluateAutoModelParameters();
```
if your goal is to make a small optimization by avoiding multiple evaluation of `Boolean.parseBoolean`, I would make `evalAutoParamsEnabled` an instance param, not a class param.

Having constructor modiying a static field is a pretty ugly and confusing pattern that should be avoided, especially if this value is then read by all instances (the only exception would be counters maybe, and even, there are cleaner way to do it).

Please change:
```   public static boolean evalAutoParamsEnabled = true; ```
into a private instance field
```    private boolean evalAutoParamsEnabled = true; ```

koniecsveta replied:
hm yes, I agree, thx

sebhrusen commented on h2o-core/src/main/java/water/api/FeatureFlagHandler.java:
```diff
@@ -0,0 +1,26 @@
+    new MRTask() {
+      @Override
+      protected void setupLocal() {
+        if (!(String.valueOf(request.value)).equals(System.getProperty("sys.ai.h2o." + request.property))) {
```
please use existing constants (here `H2O.OptArgs.SYSTEM_PROP_PREFIX`) whenever possible.

sebhrusen commented on h2o-core/src/main/java/water/api/FeatureFlagHandler.java:
```diff
@@ -0,0 +1,26 @@
+  @SuppressWarnings("unused") // called via reflection
+  public FeatureFlagV3 setFeatureFlag(int version, FeatureFlagV3 request) throws Exception {
+    Val val = Rapids.exec("(setproperty 'sys.ai.h2o." + request.property + "' '" + request.value + "')");
+    new MRTask() {
```
do you really need this task? It seems to be just verifying that `Rapids.exec` did its job, so I find this additional precaution a bit surprising.

sebhrusen commented on h2o-bindings/bin/bindings.py:
```diff
@@ -368,15 +368,19 @@ def find_field(fields, field_name):
+                    if len(gen_type) > 1:
+                        gen_type.pop(0) 
+                    field["schema_name"] = gen_type[0]
+                    mapped_fields[field["name"]] = generic_index[gen_type[0]]
```
you changed the logic here, didn't you?
You used to `pop` the first element after reading it from `gen_type` (which made sense).
Now you remove it, and then read the new head element... so you'll never read the original head: looks like a bug.

sebhrusen replied:
actually I realize that I don't understand why you had to change anything in this file, maybe it was necessary in previous phase of your PR, but I don't see the need now.
If we look at the mapping defined at lines 363-366, it hasn't changed, so the old `generic_map` should still work... it use to produce sth like:
```
dict(ModelParametersSchemaV3="P", ModelOutputSchemaV3="O")
```
and now it produces:
```
dict(ModelParametersSchemaV3=["P"], ModelOutputSchemaV3=["O"])
```
No benefit, and the `len(gen_type)` will never be > 1

koniecsveta replied:
Yes, it's on the wrong place here.. regarding whole generic map I think after this commit https://github.com/h2oai/h2o-3/pull/4361/commits/f85be8d5a4af114d2f61fb8b3d99df3170c92868 any changes in bindings.py are not needed anymore, so will revert this

michalkurka commented on h2o-core/src/main/java/hex/Model.java:
```diff
@@ -1060,7 +1080,7 @@ public ScoringInfo last_scored() {
   // Lower is better
   public float loss() {
-    switch (_parms._stopping_metric) {
+    switch (Optional.ofNullable(_parms._stopping_metric).orElse(ScoreKeeper.StoppingMetric.AUTO)) {
```
@koniecsveta why do we need to handle the null case now? What behavior changed?

koniecsveta replied:
in case when stopping_metric = AUTO and stopping_rounds = 0, then it gets evaluated as stopping_metric = null

michalkurka replied:
Interesting :) I think the behavior is somewhat strange then. We start with AUTO to turn it into null and then implicitly convert to AUTO?

Maybe it is because the parameter `_stopping_metric` is misused here (which is not fault of your PR)

michalkurka replied:
I think your change makes sense it just looks a little suspicious. Not sure how to improve it though.

honzasterba commented on h2o-core/src/main/java/hex/Model.java:
```diff
@@ -77,6 +78,18 @@ public boolean filter(KeySnapshot.KeyInfo k) {
+  public static boolean evaluateAutoModelParameters() {
+    return Boolean.parseBoolean(System.getProperty(EVAL_AUTO_PARAMS_ENABLED, "true"));
+  }
+
```
knit: whitespace

honzasterba commented on h2o-core/src/main/java/water/api/FeatureFlagHandler.java:
```diff
@@ -0,0 +1,19 @@
+import water.rapids.Val;
+
+
+public class FeatureFlagHandler extends Handler {
```
unnecessary: there is already an Rapids API that can modify system properties

michalkurka replied:
I would agree, the function on the client can evaluate the rapid expression, it doesn't have to be delegated to the backend.

honzasterba replied:
@michalkurka I am not sure what you mean, what I had in mind is

`h2o.rapids("(setproperty \"{}\" \"{}\")".format("sys.ai.h2o.name", "value"))`

michalkurka replied:
Yes

sebhrusen replied:
@michalkurka , @honzasterba 
do we really want to completely remove the client function that @koniecsveta previously created (it's been removed after your comments)?
1. No one knows how to use this `rapids` function.
2. No one knows the system properties that we allow them to set.
3. the `set_feature_flag` was self documented.
4. that function was also limited to the flags we want users to be aware of, preventing typos (if user makes a typo in a system property with rapids, he can't notice, it will just create the sys property), or to change a property that is not supposed to be changed...
5. Feature flags are also a very good strategy to rollout experimental features, so this client could be profitable in the future.

honzasterba replied:
I dont think modifying feature flags at runtime is something we want to expose/support, I dont think most of the code is designed to work with them changing at runtime, thats why I would stick to the Rapids API for testing.

sebhrusen replied:
>  I dont think modifying feature flags at runtime is something we want to expose/support.

Well, I personally wish the opposite. If we design sth in such a way that it can be enabled/disabled, it should be exposed rather than frustrate the customer wanting to try or disable that feature.

>  I dont think most of the code is designed to work with them changing at runtime

If it can't be changed at runtime, this is not a feature flag. That's why the function was narrowed to a specific list of features. (I know at least 1 more currently: AutoML XGB on multinode...).

sebhrusen replied:
Personally I'm fine with current state of the code, I was even suggesting @koniecsveta  to implement the flag API in a separate PR. Just want to know if you think it's worth to create a separate PR.

honzasterba commented on h2o-core/src/main/java/water/api/SchemaMetadata.java:
```diff
@@ -203,8 +204,9 @@ public static FieldMetadata createIfApiAnnotation(Schema schema, Field f, List<F
         return new FieldMetadata(schema, f, superclassFields);
-
-      Log.warn("Skipping field that lacks an annotation: " + schema.toString() + "." + f);
+      if (!(Modifier.isPrivate(f.getModifiers()))) { // don't warn in case of private field without annotation
```
why is this necessary?

koniecsveta replied:
it is because of ModelSchemaV3.java#input_parameters field. It is a private field and it doesn't have to be exposed in the API. It is used only in `ModelParametersSchemaV3.writeParametersJSON(ab, parameters, input_parameters, defaults ,"parameters");` But when I just remove @API notation, INFO check starts failing, because of the field without annotation found. So I can either expose what is not needed to be exposed or modify this check to not warn on fields without annotation when they are private. Private fields are also pretty rare in schemas, we have some only in TimelineV3. These have api annotation, so they won't be affected by this change. Do you think it would be better to expose and leave this check as it was?

honzasterba replied:
Those private fields in TimelineV3 would be affected, because if their annotation was accidentally removed no warning message would be logged. 
I am having hard time coming up with a good solution, only thing that comes to mind is to have some @Ignore annotation, but thats also not very pretty.
Another thought you say it is not exposed in the API, yet in writeParametersJSON you write out the value into the JSON so in fact it is exposed in the API and should be public, problem solved.

koniecsveta replied:
sounds good public

sebhrusen replied:
@honzasterba do you know why those Timelinev3 fields are even private? if they're part of the API? 
I don't think we should make things more complex only because of an old decision (private fields that should be public) that we don't understand or that doesn't  seem to make sense.
Is it not this class that should have all its API fields public?

Making `input_parameters` public (with an API annotation) is also nonsense , because it is not exposed as such, but only as the `input` field in for each param, and we don't want to expose those values twice.

honzasterba replied:
@sebhrusen  Sorry I have not noticed this is about a field on ModelSchemaV3, you are correct

AFAIK schema does not differentiate between public/private so we could fix TimelineV3 and fix this check to only apply to non-anotated public fields and also make Schema only check public fields

sebhrusen replied:
Yes, for consistency, I'd also fix `TimelineV3` to use public fields. I think this is the only use case in the code base.
Did see any other in `core`, `algos` or `automl`.

koniecsveta replied:
So... will leave ModelSchemaV3.java and also binding.py as they are, so that we have input_parameters only in context where we want them to be. And change private fields in TimelineV3 to public so that if their annotation was accidentally removed we will get warning. We have also some protected fields and fields without explicit modifier.. So I would not warn only on private non-anotated fields and warn on all the others non-anotated (as it is now in 02884ea) Allright?

michalkurka commented on h2o-core/src/main/java/water/api/schemas3/ModelParametersSchemaV3.java:
```diff
@@ -63,6 +63,10 @@
 
+  @API(level = API.Level.expert, direction = API.Direction.INOUT,
+          help = "Whether to evaluate input parameters of value AUTO.")
+  public boolean evaluate_auto;
```
@koniecsveta did you decide to keep the evaluate auto parameter? Can if so - can you please get a feedback in #h2o-3-api-updates?

sebhrusen replied:
interesting, I thought this one got removed as it doesn't appear in the clients anymore

michalkurka commented on h2o-core/src/main/java/water/api/schemas3/ModelSchemaV3.java:
```diff
@@ -43,6 +43,8 @@
   @API(help="Checksum for all the things that go into building the Model.", direction=API.Direction.OUTPUT)
   protected long checksum;
 
+  private PS input_parameters;
```
@koniecsveta Why is this field private-only? It is surprising to see something like that in a schema class.

sebhrusen replied:
It's due to the unique way parameters are serialized to JSON in this class: so this private field is used to transfer the input value from the model instance to the final JSON.

cf. also conversation in `SchemaMetadata`

personally, I don't have any problem with private fields in schemas in that case.
I'm much more confused by private fields exposed as API in `TimelineV3`, which looks like nonsense to me.

honzasterba commented on h2o-py/tests/pyunit_utils/utilsPY.py:
```diff
@@ -3219,7 +3219,13 @@ def model_seed_sorted(model_list):
             if model_list.models[index]._model_json["parameters"][pIndex]["name"]=="seed":
                 model_seed_list[index]=model_list.models[index]._model_json["parameters"][pIndex]["actual_value"]
                 break
+    print("model seed list before sort: ")
```
these look like debugging printouts, do we want to keep those?

koniecsveta replied:
no, we don't, there is a separate JIRA for this issue. will remove this

honzasterba commented on h2o-py/tests/testdir_algos/automl/pyunit_automl_effective_parameters.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def test_actual_default_input_stopping_rounds():
+    h2o.rapids("(setproperty \"{}\" \"{}\")".format("sys.ai.h2o.algos.evaluate_auto_model_parameters", "true"))
```
this is unnecessary, since that is the default value

honzasterba commented on h2o-py/tests/testdir_algos/gbm/pyunit_effective_parameters_gbm.py:
```diff
@@ -0,0 +1,104 @@
+    assert gbm1.parms['categorical_encoding']['input_value'] == 'AUTO'
+    assert gbm1.parms['categorical_encoding']['actual_value'] == 'AUTO'
+
+    h2o.rapids("(setproperty \"{}\" \"{}\")".format("sys.ai.h2o.algos.evaluate_auto_model_parameters", "true"))
```
I think this should be in a finally block since it might break other tests otherwise

honzasterba commented on h2o-r/tests/testdir_algos/gbm/runit_GBM_effective_parameters.R:
```diff
@@ -0,0 +1,27 @@
+source("../../../scripts/h2o-r-test-setup.R")
+
+test.gbm.effective_parameters <- function() {
+    h2o.rapids('(setproperty "sys.ai.h2o.algos.evaluate_auto_model_parameters" "true")')
```
unnecessary
